### PR TITLE
Use Fractal to organise and document components

### PR DESCRIPTION
**Pull request by tyom**
Reorganise components to use with Clearleft's [Fractal](http://fractal.build/).

This includes basic global CSS components that can be used on other NHS.UK sites. The static [documentation site](https://nhsuk.github.io/frontend/0.1.0/) is also generated.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,23 +1,107 @@
...
+
+Configure Nunjucks to add this package's template path to your Nunjucks environment:
+
+ ```js
+ const path = require('path');
+const nunjucksEnv = nunjucks.configure([
+  path.resolve(__dirname, '/app/views'),
+  path.resolve(__dirname, '/node_modules/nhsuk-frontend/src/templates')
+], {
+  autoescape: true
```
Comment (r4vi) [inline README.md:82]: this is the default: (from nunjucks/src/environment.js)

    // The autoescape flag sets global autoescaping. If true,
    // every string variable will be escaped by default.
    // If false, strings can be manually escaped using the `escape` filter.
    // defaults to true
    this.opts.autoescape = opts.autoescape != null ? opts.autoescape : true;

Comment (tyom) [reply README.md:82]: I added it there just as an example to show that following the array of paths is the configuration object.

File: README.md [hunk 2]
```diff
@@ -1,23 +1,107 @@
...
+**Nunjucks blocks**
+
+- `head` - contains the whole `head` element
+  - `head_title_content` - contains `title` element content
+  - `head_content` - wraps main content for `head` element
+  - `head_stylesheets` - wraps stylesheet declarations
+- `body` - contains the whole `body` element
+  - `body_skiplinks` - wraps container with "skip to content" link (first element inside `body`)
+  - `body_notifications` - wraps cookie message container (above `site_header`, after `skiplinks`)
+  - `body_site_header` - wraps side header
```
Comment (teneightfive) [inline README.md]: Typo - _side_ should be _site_

File: src/scss/components/01-global-header/_global-header.scss [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+.global-header__link {
+  background: url("../../assets/images/logotype-nhs-colour__reverse.png") no-repeat;
+  background-size: contain;
+  display: block;
+  height: 32px;
+  text-indent: -999em;
+  width: 79px;
+
+  @media(min-resolution: 192dpi) {
```
Comment (teneightfive) [inline src/scss/components/01-global-header/_global-header.scss]: Does autoprefixer handle the variations for this that we currently include in a mixin?

Comment (tyom) [reply src/scss/components/01-global-header/_global-header.scss]: Autoprefixer gives a pretty good range which seemingly falls into what we specify in the config. Looking at caniuse.com browser support for [this feature](http://caniuse.com/#feat=css-media-resolution) I can see that `-webkit` and unprefixed version covers most of the use cases (IE9+, FF16+, Opera 12.1+, Chrome etc).

I don't see any value in adding very edge case prefixes. Only to increase complexity by having to use and maintain mixins.

**Comment by teneightfive** [comment]
Is the purpose of the number prefix on component folders to determine display order? Not sure I like the idea of it being determined by the folder name. If we want to re-order in the future or add a component in between item 5 and 6 and we have 15 components that change would create an unnecessarily large diff. 

Is there a problem with components being alphabetically ordered?

**Comment by teneightfive** [comment]
We should also consider what components we actually want to offer with this package. It could be useful to break down things like the global header and footer into a component but there might not be a need for this to be displayed in the living styleguide if they come as part of the site chrome.

If we think of this as the things needed to create assemblies that have a consistent look across NHS.UK the template could be a separate part to the components that are used inside that layout.

**Comment by tyom** [comment]
> Is the purpose of the number prefix on component folders to determine display order? Not sure I like the idea of it being determined by the folder name. If we want to re-order in the future or add a component in between item 5 and 6 and we have 15 components that change would create an unnecessarily large diff.

I agree with you. I soon realised it would be painful to reorder. I think there's a way to order components from meta data. Alphabetical is fine, I guess, but it helps to group certain things.

I'll try using meta data for this.

**Comment by tyom** [comment]
> We should also consider what components we actually want to offer with this package. It could be useful to break down things like the global header and footer into a component but there might not be a need for this to be displayed in the living styleguide if they come as part of the site chrome.

I don't have a problem documenting global level components. But instead of breaking them up we could document the whole base template assembly.

**Comment by teneightfive** [comment]
>I don't have a problem documenting global level components. But instead of breaking them up we could document the whole base template assembly.

Yep, that sounds like a sensible start. Reduces the noise in the living styleguide around components. I think a need of "I need the header and footer components..." would be met by the base template so we can remove them from the component list.

File: README.md [hunk 4]
```diff
@@ -1,23 +1,107 @@
...
+
+```bash
+npm run build
 ```
 
 Build docs for deployment:
 
 ```bash
-npm run docs:build
+npm run fractal:build
```
Comment (teneightfive) [inline README.md]: The task has been renamed back to `docs:build` so needs replicating here.

File: fractal.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    markdownInline(str) {
+      return md.renderInline(str)
+    },
+    slugify(str) {
+      return str.toLowerCase().replace(/[^\w]+/g, '')
+    },
+    stringify() {
+      return JSON.stringify(this, null, 2)
+    },
+  }
```
Comment (teneightfive) [inline fractal.js]: Trailing comma consistency?

File: fractal.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+fractal.components.set('path', `${paths.src}/scss/components`)
+
+// Docs config
+fractal.docs.engine(nunjucks)
+fractal.docs.set('ext', '.md')
+fractal.docs.set('path', `${paths.src}/docs`)
+
+// Web UI config
+fractal.web.theme(nhsCustomTheme)
+fractal.web.set('static.path', paths.dist);
```
Comment (teneightfive) [inline fractal.js]: Remove semi-colons for consistency?

File: fractal/theme/index.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  styles: ['default', '/themes/nhsuk/theme.css'],
+  nav: ['docs', 'components'],
+  panels: ['html', 'context', 'resources', 'info']
+})
+
+customTheme.addStatic(path.join(__dirname, 'assets'), 'themes/nhsuk')
+
+customTheme.addLoadPath(path.join(__dirname, 'views'))
+
+module.exports = customTheme;
```
Comment (teneightfive) [inline fractal/theme/index.js]: No need for semi-colon

**Review by teneightfive — CHANGES_REQUESTED** [review]

**Comment by teneightfive** [comment]
Looks really good as a start for maintaining and sharing layouts and components.

**Review by teneightfive — APPROVED** [review]

**Pull request merged**